# Análise dos dados das variáveis relacionadas ao IDEB

O notebook contém experimentos com essas variáveis. São utilizadas duas bases de dados:<br>
- Váriáveis derivadas do SAEB (informações socioeconômicas)<br> 
- Variáveis com informações das escolas e professores<br>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
from sklearn.feature_selection import SelectKBest, RFE, RFECV, SelectFromModel
from sklearn.feature_selection import mutual_info_classif, f_classif, chi2
from sklearn.preprocessing import LabelEncoder
import numpy as np
from numpy import array 
from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import sklearn.metrics as metrics

## Importando bases de dados

In [ ]:
#Importando variáveis derivadas do SAEB

#base_estudantes_se = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/variaveis_derivadas_se.csv')

base_estudantes_se = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_se.csv')

#base_estudantes_se.set_index('ID_ESCOLA', inplace=True)

#base_estudantes_al = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/variaveis_derivadas_al.csv')

#base_estudantes_al = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_al.csv')

#base_estudantes_al.set_index('ID_ESCOLA', inplace=True)

#base_estudantes_ba = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_ba.csv')

dados_pessoais = base_estudantes_se

#base_estudantes_ba.set_index('ID_ESCOLA', inplace=True)

#dados_pessoais = pd.merge(base_estudantes_se, base_estudantes_al, how = 'outer')

#dados_pessoais = pd.merge(base_estudantes_ba, dados_pessoais, how = 'outer')

In [ ]:
dados_pessoais.set_index('ID_ESCOLA', inplace=True)

In [ ]:
dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
28012631              2017          2  ...               2               1
61178208              2017          2  ...               5               2
61211437              2017          2  ...              10               7
28017846              2017          2  ...               3               1
28017854              2017          2  ...               5              10
...                    ...        ...  ...             ...             ...
61247893              2017          2  ...               3               5
61229418              2017          2  ...               4               5
61214852              2017          2  ...               2               3
61215477              2017          2  ...               2               1
61231744              2017          2  ...               2               2

[195 rows x 234 columns]

**Pré-processamento da base de dados pessoais**

In [ ]:
#Função que verifica se o ID de uma escola é inválido
#Segundo o Dicionário de Dados da base o SAEB 6 - Os códigos dos Municípios e 
#das Escolas que começam com o dígito "6" são máscaras, isto é, são códigos fictícios.
def retira_id_ficticio(id_escola):
    if id_escola[0] == "6":
        return True
    else:
        return False

In [ ]:
#Se retirou as escolas fictícias
for escola in dados_pessoais.index:
    if (retira_id_ficticio(str(escola))):
        dados_pessoais.drop(escola, inplace=True)
        

dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
28012631              2017          2  ...               2               1
28017846              2017          2  ...               3               1
28017854              2017          2  ...               5              10
28018400              2017          2  ...               6               7
28018435              2017          2  ...               7              18
...                    ...        ...  ...             ...             ...
28009398              2017          2  ...              22              29
28009487              2017          2  ...               4               4
28014839              2017          2  ...               8               5
28010132              2017          2  ...              15               6
28028562              2017          2  ...               9              12

[77 rows x 234 columns]

In [ ]:
#Importando variáveis com informações das escolas e professores
#base_escola_se = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-SE.csv')

base_escola_se = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-SE.csv')

base_escola_se.set_index('CO_ENTIDADE', inplace=True)

#base_escola_al = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-AL.csv')

#base_escola_al = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-AL.csv')

#base_escola_al.set_index('CO_ENTIDADE', inplace=True)

#base_escola_ba = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-BA.csv')

#base_escola_ba = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-BA.csv')

#base_escola_ba.rename(columns={'Unnamed: 0': 'CO_ENTIDADE'}, inplace=True)

#base_escola_ba.set_index('CO_ENTIDADE', inplace=True)

#dados_escola = pd.merge(base_escola_se, base_escola_al, how = 'outer')

#dados_escola = pd.concat([dados_escola, base_escola_ba], join='outer')

dados_escola = base_escola_se

In [ ]:
#dados_escola.set_index('CO_ENTIDADE', inplace=True)

dados_escola

NU_ANO_CENSO  ... Ideb-2017
CO_ENTIDADE                ...          
28017943             2017  ...       6.9
28010124             2017  ...       6.6
28018052             2017  ...       6.4
28032497             2017  ...       6.1
28017773             2017  ...       5.8
...                   ...  ...       ...
28022033             2017  ...       4.0
28005759             2017  ...       3.0
28018940             2017  ...       3.0
28025903             2017  ...       3.0
28029577             2017  ...       3.0

[96 rows x 185 columns]

In [ ]:
col_ideb = dados_escola['Ideb-2017']

dados_escola.drop('Ideb-2017', axis=1, inplace=True)

In [ ]:
col_ideb

CO_ENTIDADE
28017943    6.9
28010124    6.6
28018052    6.4
28032497    6.1
28017773    5.8
           ... 
28022033    4.0
28005759    3.0
28018940    3.0
28025903    3.0
28029577    3.0
Name: Ideb-2017, Length: 96, dtype: float64

In [ ]:
dados_escola

NU_ANO_CENSO  ... RelAlunProf
CO_ENTIDADE                ...            
28017943             2017  ...           7
28010124             2017  ...           4
28018052             2017  ...           5
28032497             2017  ...           1
28017773             2017  ...           6
...                   ...  ...         ...
28022033             2017  ...          12
28005759             2017  ...           5
28018940             2017  ...           8
28025903             2017  ...          11
28029577             2017  ...          17

[96 rows x 184 columns]

#### Pré-processamento da base das escolas

In [ ]:
#Retirando valores ausentes
filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

# retorna o número de valores ausentes para cada coluna
dados_escola[listOfColumnNames].isnull().sum()

CO_ORGAO_REGIONAL                 1
TP_CATEGORIA_ESCOLA_PRIVADA      76
IN_CONVENIADA_PP                 76
TP_CONVENIO_PODER_PUBLICO        96
IN_MANT_ESCOLA_PRIVADA_EMP       76
IN_MANT_ESCOLA_PRIVADA_ONG       76
IN_MANT_ESCOLA_PRIVADA_SIND      76
IN_MANT_ESCOLA_PRIVADA_SIST_S    76
IN_MANT_ESCOLA_PRIVADA_S_FINS    76
CO_ESCOLA_SEDE_VINCULADA         96
CO_IES_OFERTANTE                 96
TP_OCUPACAO_PREDIO_ESCOLAR        1
TP_OCUPACAO_GALPAO               96
IN_PREDIO_COMPARTILHADO           1
IN_FUNDAMENTAL_CICLOS             3
TP_INDIGENA_LINGUA               96
CO_LINGUA_INDIGENA               96
dtype: int64

In [ ]:
#Retirando colunas que apresentam mais de 60% do valores nulos
dados_escola.dropna(thresh=len(dados_escola)*0.6, axis=1, inplace=True)

#Preenchendo valores ausentes
dados_escola.fillna(0, inplace=True)

filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

# retorna o número de valores ausentes para cada coluna
dados_escola[listOfColumnNames].isnull().sum()

CO_ORGAO_REGIONAL             0
TP_OCUPACAO_PREDIO_ESCOLAR    0
IN_PREDIO_COMPARTILHADO       0
IN_FUNDAMENTAL_CICLOS         0
dtype: int64

In [ ]:
dados_escola[listOfColumnNames] = dados_escola[listOfColumnNames].astype(int)

In [ ]:
dados_escola.insert(len(dados_escola.columns), 'Ideb-2017', col_ideb)
#['Ideb-2017'] = col_ideb

In [ ]:
dados_escola

NU_ANO_CENSO  ... Ideb-2017
CO_ENTIDADE                ...          
28017943             2017  ...       6.9
28010124             2017  ...       6.6
28018052             2017  ...       6.4
28032497             2017  ...       6.1
28017773             2017  ...       5.8
...                   ...  ...       ...
28022033             2017  ...       4.0
28005759             2017  ...       3.0
28018940             2017  ...       3.0
28025903             2017  ...       3.0
28029577             2017  ...       3.0

[96 rows x 172 columns]

In [ ]:
#Colunas categóricas
filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.object]

listOfColumnNames = list(filteredColumns.index)

dados_escola[listOfColumnNames]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  


NO_ENTIDADE  ... DT_ANO_LETIVO_TERMINO
CO_ENTIDADE                                                     ...                      
28017943                                  COLEGIO AMADEUS LTDA  ...            24/11/2017
28010124     ESCOLA DE EDUCACAO BASICA MONSENHOR JOSE DE SO...  ...            08/12/2017
28018052                      COLEGIO SANTANNA DE SERGIPE LTDA  ...            01/12/2017
28032497                                      COLEGIO NOTA DEZ  ...            15/12/2017
28017773                                 COL AMERICANO BATISTA  ...            01/12/2017
...                                                        ...  ...                   ...
28022033           COLEGIO ESTADUAL CLEONICE SOARES DA FONSECA  ...            21/12/2017
28005759                   COLEGIO ESTADUAL EMILIANO GUIMARAES  ...            16/01/2018
28018940                        COLEGIO ESTADUAL SANTOS DUMONT  ...            21/03/2018
28025903                  COLEGIO ESTADUAL COMENDADOR CALAZANS  ...            05/02/2018
28029577     COLEGIO ESTADUAL PREFEITO JOALDO LIMA DE CARVALHO  ...            29/01/2018

[96 rows x 3 columns]

In [ ]:
#Tratando datas, convertendo para timestamp

inicio = pd.to_datetime(dados_escola['DT_ANO_LETIVO_INICIO'], infer_datetime_format = True)
termino = pd.to_datetime(dados_escola['DT_ANO_LETIVO_TERMINO'], infer_datetime_format = True)

dados_escola['DT_ANO_LETIVO_INICIO'] = inicio.values.astype (np.int64)
dados_escola['DT_ANO_LETIVO_TERMINO'] = termino.values.astype (np.int64)

In [ ]:
dados_escola.drop('NO_ENTIDADE', axis=1, inplace=True)

In [ ]:
dados_escola.columns.to_list()

['NU_ANO_CENSO',
 'CO_ORGAO_REGIONAL',
 'TP_SITUACAO_FUNCIONAMENTO',
 'DT_ANO_LETIVO_INICIO',
 'DT_ANO_LETIVO_TERMINO',
 'CO_REGIAO',
 'CO_MESORREGIAO',
 'CO_MICRORREGIAO',
 'CO_UF',
 'CO_MUNICIPIO',
 'CO_DISTRITO',
 'TP_DEPENDENCIA',
 'TP_LOCALIZACAO',
 'TP_REGULAMENTACAO',
 'IN_LOCAL_FUNC_PREDIO_ESCOLAR',
 'TP_OCUPACAO_PREDIO_ESCOLAR',
 'IN_LOCAL_FUNC_SALAS_EMPRESA',
 'IN_LOCAL_FUNC_SOCIOEDUCATIVO',
 'IN_LOCAL_FUNC_UNID_PRISIONAL',
 'IN_LOCAL_FUNC_PRISIONAL_SOCIO',
 'IN_LOCAL_FUNC_TEMPLO_IGREJA',
 'IN_LOCAL_FUNC_CASA_PROFESSOR',
 'IN_LOCAL_FUNC_GALPAO',
 'IN_LOCAL_FUNC_SALAS_OUTRA_ESC',
 'IN_LOCAL_FUNC_OUTROS',
 'IN_PREDIO_COMPARTILHADO',
 'IN_AGUA_FILTRADA',
 'IN_AGUA_REDE_PUBLICA',
 'IN_AGUA_POCO_ARTESIANO',
 'IN_AGUA_CACIMBA',
 'IN_AGUA_FONTE_RIO',
 'IN_AGUA_INEXISTENTE',
 'IN_ENERGIA_REDE_PUBLICA',
 'IN_ENERGIA_GERADOR',
 'IN_ENERGIA_OUTROS',
 'IN_ENERGIA_INEXISTENTE',
 'IN_ESGOTO_REDE_PUBLICA',
 'IN_ESGOTO_FOSSA',
 'IN_ESGOTO_INEXISTENTE',
 'IN_LIXO_COLETA_PERIODICA',
 'IN_LIXO_

In [ ]:
dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
28012631              2017          2  ...               2               1
28017846              2017          2  ...               3               1
28017854              2017          2  ...               5              10
28018400              2017          2  ...               6               7
28018435              2017          2  ...               7              18
...                    ...        ...  ...             ...             ...
28009398              2017          2  ...              22              29
28009487              2017          2  ...               4               4
28014839              2017          2  ...               8               5
28010132              2017          2  ...              15               6
28028562              2017          2  ...               9              12

[77 rows x 234 columns]

### Unindo as bases de dados 

Foi realizado join nos dois datasets, nesse momento se percebeu que não existem variáveis escolas particulares na base derivada do SAEB

In [ ]:
dataset = dados_pessoais.join(dados_escola, how="inner")

In [ ]:
dataset

ID_PROVA_BRASIL  ID_REGIAO  ...  RelAlunProf  Ideb-2017
28012631             2017          2  ...            2        3.1
28017846             2017          2  ...            5        3.9
28017854             2017          2  ...            9        2.9
28018400             2017          2  ...           11        3.9
28018435             2017          2  ...           11        3.7
...                   ...        ...  ...          ...        ...
28009398             2017          2  ...           11        3.5
28009487             2017          2  ...           13        3.9
28014839             2017          2  ...           11        3.5
28010132             2017          2  ...           16        3.5
28028562             2017          2  ...            7        3.5

[76 rows x 405 columns]

In [ ]:
dados_escola.dtypes[dados_escola.dtypes == np.float64]

Ideb-2017    float64
dtype: object

In [ ]:
dataset['ID_DEPENDENCIA_ADM'].unique()

array([2, 1])

In [ ]:
#dataset.to_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa Ideb/variaveis_derivadas_escola_estudante_5.0.csv')
dataset.to_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_escola_estudante_5.0.csv')

### Discretização dos rótulos, nota do Ideb

**Discretização:** É o processo de transferência de funções contínuas, modelos, variáveis e equações em contrapartes discretas.

In [ ]:
# São três classes média, acima da média e abaixo da média
def discretizacao(ideb, mean_ideb):
    classe = []
    for nota in ideb['Ideb-2017']:
        if (mean_ideb == nota):
            classe.append('média')
        elif (mean_ideb < nota):
            classe.append('acima da média')
        else:
            classe.append('abaixo da média')
    return classe

In [ ]:
#discretizando as notas do Ideb

Ideb = dataset[['Ideb-2017']]

mean_Ideb = round(Ideb.mean(),1)

classe_ideb = discretizacao(Ideb, mean_Ideb['Ideb-2017'])

Ideb.insert(1, 'classe-Ideb', classe_ideb)

Ideb

Ideb-2017      classe-Ideb
28012631        3.1  abaixo da média
28017846        3.9   acima da média
28017854        2.9  abaixo da média
28018400        3.9   acima da média
28018435        3.7   acima da média
...             ...              ...
28009398        3.5   acima da média
28009487        3.9   acima da média
28014839        3.5   acima da média
28010132        3.5   acima da média
28028562        3.5   acima da média

[76 rows x 2 columns]

In [ ]:
#Variavel a ser prevista
y = Ideb['classe-Ideb']

**Seleção de Atributos:**  É um passo da fase de pré-processamento que tem a finalidade de selecionar o melhor subconjunto de atributos que pode substituir o conjunto de dados original. 

https://www.analyticsvidhya.com/blog/2016/12/introduction-to-feature-selection-methods-with-an-example-or-how-to-select-the-right-variables/#:~:text=Embedded%20methods%20combine%20the%20qualities,penalization%20functions%20to%20reduce%20overfitting.

Abordagens:

- Filtro: a relevância dos atributos é avaliada independentemente – não
envolve o uso de um método de AM ;

- Wrapper (Envólucro): Avalia o subconjunto de atributos arti culado ao método de AM que realizará a indução do conceito;

- Embedded (Embutidos): Os métodos incorporados combinam as qualidades dos métodos de filtro e wrapper. Ele é implementado por algoritmos que possuem seus próprios métodos de seleção de recursos integrados.

In [ ]:
#Filter based
# f_classif: ANOVA F-value between label/feature for classification tasks.
#ANOVA stands for Analysis of variance. It is similar to LDA except for the fact that it is operated using one or more categorical 
#independent features and one continuous dependent feature. It provides a statistical test 
#of whether the means of several groups are equal or not.
def selecao_atributos(X, y, quant):
    select = SelectKBest(score_func=f_classif, k=quant)
    #select = SelectKBest(score_func=chi2, k=quant)
    z = select.fit_transform(X, y)
    
    filter = select.get_support(indices=True)
    features = array(X.columns)
    
    return X.iloc[:, filter]

In [ ]:
#Wrapper-based
#from sklearn.cluster import KMeans

def wrapper_based(X, X_transform, y, quant):

  #select_features = RFE(estimator=KMeans(n_clusters=3), n_features_to_select=100)
  select_features = RFE(estimator=SVC(kernel="linear"), n_features_to_select=100)
  '''select_features = RFECV(
      estimator=SVC(kernel="linear"),
      min_features_to_select= quant,
      cv = 5
  )'''
  select_features.fit(X_transform, y)

  filter = select_features.get_support(indices=True)
  return X.iloc[:, filter]

In [ ]:
# Embedded-based

def embedded_based(X, X_transform, y, quant):

  select_features = SelectFromModel(LogisticRegression(), max_features=quant)
  #select_features = SelectFromModel(DecisionTreeClassifier(), max_features=quant)
  #select_features = SelectFromModel(SVC(kernel="rbf"), max_features=quant)
  #select_features = SelectFromModel(RandomForestClassifier(), max_features=quant)
  select_features.fit(X_transform, y)

  filter = select_features.get_support(indices=True)

  return X.iloc[:, filter]

**Métricas de Avaliação**


In [ ]:
def acuracia(resultados, classes):
    #resultados = cross_val_predict(modelo, textos, classes, cv=10)  
    return 'Acurácia do modelo: {} %'.format(round(metrics.accuracy_score(classes, resultados)*100, 2))

**Normalização:** é mudar os valores das colunas numéricas no conjunto de dados para usar uma escala comum, sem distorcer as diferenças nos intervalos de valores nem perder informações.

- **Normalização MinMax:** Para cada recurso, o valor mínimo desse recurso é transformado em 0, o valor máximo é transformado em 1 e todos os outros valores são transformados em um decimal entre 0 e 1.

In [ ]:
def minmax(data):
    scaler = MinMaxScaler()
    data_scaler = scaler.fit_transform(data)
    return data_scaler

## Criação dos modelos

**Serão feitas três análises:** Apenas com as variáveis socioeconômicas, com as variáveis de escolas e professores e depois com o dataset completo

Para cada uma serão criados modelos com todos os atributos e também com seleção de atributos, a fim de verificar o de maior acurácia

In [ ]:
def inicializar_modelos():
    naive = MultinomialNB()
    random = RandomForestClassifier(n_estimators=150)
    decision_tree = DecisionTreeClassifier()
    svm = SVC(kernel="linear")
    r_logistic = LogisticRegression(max_iter=250)
    knn = KNeighborsClassifier(n_neighbors=3)
    return naive, random, decision_tree, svm, r_logistic, knn

#### -Primeira análise: variáveis socioeconômicas derivadas da prova SAEB**

In [ ]:
#Foi retirada colunas do 'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF'
columns = dataset.columns[3:234]

columns

Index(['ID_MUNICIPIO', 'ID_AREA', 'ID_DEPENDENCIA_ADM', 'ID_LOCALIZACAO',
       'TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q020_A', 'TX_RESP_Q020_B',
       'TX_RESP_Q021_A', 'TX_RESP_Q021_B',
       ...
       'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 'TX_RESP_Q003_E',
       'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I',
       'TX_RESP_Q003_J', 'TX_RESP_Q003_L'],
      dtype='object', length=231)

In [ ]:
X =  dataset[columns]

In [ ]:
#Questões de sexo, raça, ano, mes de nascimento, conformidade com censo
question = ['TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D', 
            'TX_RESP_Q004_E', 'TX_RESP_Q004_F', 'TX_RESP_Q004_G', 'TX_RESP_Q004_H']

#delete_at = ['IN_SITUACAO_CENSO']

X.drop(question, axis=1, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
filteredColumns = X.dtypes[X.dtypes == np.float]

listOfColumnNames = list(filteredColumns.index)

X[listOfColumnNames]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: []
Index: [28012631, 28017846, 28017854, 28018400, 28018435, 28018478, 28018494, 28018508, 28018516, 28018761, 28018818, 28018915, 28018940, 28031911, 28032578, 28006542, 28022033, 28006720, 28006739, 28012836, 28005651, 28014693, 28024532, 28000897, 28027736, 28000951, 28013093, 28013158, 28025326, 28007166, 28029577, 28029569, 28011074, 28011082, 28032225, 28007867, 28005759, 28008227, 28008235, 28001400, 28005848, 28013247, 28004264, 28001699, 28006330, 28013514, 28020464, 28032306, 28004582, 28023277, 28015983, 28002245, 28003047, 28003055, 28026705, 28013700, 28012151, 28012542, 28004809, 28030095, 28016890, 28040201, 28025903, 28013921, 28013972, 28014812, 28021045, 28021053, 28008421, 28016106, 28006356, 28009398, 28009487, 28014839, 28010132, 28028562]

[76 rows x 0 columns]

**Divisão entre conjunto de treino e teste**

In [ ]:
X_scaler = minmax(X)

X_train, X_test, y_train, y_test = train_test_split(X_scaler, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 56.25 %
Naive Bayes: Acurácia do modelo: 68.75 %
Decision Tree: Acurácia do modelo: 62.5 %
SVM: Acurácia do modelo: 81.25 %
Logistic Regression: Acurácia do modelo: 81.25 %
KNN: Acurácia do modelo: 56.25 %


**-Primeira Análise: Aplicando Seleção de Atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X, y, 100)

X_selection

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [148] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


ID_AREA  ID_DEPENDENCIA_ADM  ...  TX_RESP_Q023_G  TX_RESP_Q003_L
28012631        2                   2  ...               7               1
28017846        1                   2  ...               9               1
28017854        1                   2  ...              24              10
28018400        1                   2  ...              12               7
28018435        1                   2  ...              29              18
...           ...                 ...  ...             ...             ...
28009398        2                   2  ...              46              29
28009487        2                   2  ...               9               4
28014839        2                   2  ...              20               5
28010132        2                   2  ...              34               6
28028562        2                   2  ...              24              12

[76 rows x 100 columns]

In [ ]:
X_scaler = minmax(X_selection)

X_train, X_test, y_train, y_test = train_test_split(X_scaler, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 81.25 %
Naive Bayes: Acurácia do modelo: 62.5 %
Decision Tree: Acurácia do modelo: 62.5 %
SVM: Acurácia do modelo: 75.0 %
Logistic Regression: Acurácia do modelo: 81.25 %
KNN: Acurácia do modelo: 75.0 %


**-Primeira Análise: Aplicando Seleção de Atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_scaler, y, 100)

X_embedded

ID_MUNICIPIO  ID_DEPENDENCIA_ADM  ...  TX_RESP_Q045_C  TX_RESP_Q051_B
28012631       2800100                   2  ...               1               0
28017846       2800308                   2  ...               4               3
28017854       2800308                   2  ...               1              14
28018400       2800308                   2  ...               2              13
28018435       2800308                   2  ...               5              19
...                ...                 ...  ...             ...             ...
28009398       2807105                   2  ...               3              39
28009487       2807105                   2  ...               0               0
28014839       2807204                   2  ...               1               2
28010132       2807402                   2  ...               2              13
28028562       2807600                   2  ...               0              12

[76 rows x 42 columns]

In [ ]:
X_scaler_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_scaler_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 68.75 %
Naive Bayes: Acurácia do modelo: 37.5 %
Decision Tree: Acurácia do modelo: 31.25 %
SVM: Acurácia do modelo: 18.75 %
Logistic Regression: Acurácia do modelo: 31.25 %
KNN: Acurácia do modelo: 75.0 %


**-Primeira Análise: Aplicando Seleção de Atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_scaler, y, 100)

X_wrapper

ID_MUNICIPIO  ID_AREA  ...  TX_RESP_Q051_A  TX_RESP_Q051_B
28012631       2800100        2  ...              18               0
28017846       2800308        1  ...              36               3
28017854       2800308        1  ...              88              14
28018400       2800308        1  ...              60              13
28018435       2800308        1  ...             120              19
...                ...      ...  ...             ...             ...
28009398       2807105        2  ...             245              39
28009487       2807105        2  ...              42               0
28014839       2807204        2  ...              69               2
28010132       2807402        2  ...             122              13
28028562       2807600        2  ...              86              12

[76 rows x 100 columns]

In [ ]:
X_scaler_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_scaler_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 68.75 %
Naive Bayes: Acurácia do modelo: 62.5 %
Decision Tree: Acurácia do modelo: 68.75 %
SVM: Acurácia do modelo: 62.5 %
Logistic Regression: Acurácia do modelo: 62.5 %
KNN: Acurácia do modelo: 75.0 %


#### -Segunda análise: Análise com o dataset completo

In [ ]:
#Lembrar de tratar datas

'''question = ['TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q003_L', 'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D', 'CO_REGIAO',
            'TP_SITUACAO_FUNCIONAMENTO', 'TP_REGULAMENTACAO',
            'CO_MESORREGIAO', 'CO_MICRORREGIAO', 'CO_DISTRITO', 'TP_LOCALIZACAO',  'CO_ORGAO_REGIONAL', 'TP_DEPENDENCIA',
            'TX_RESP_Q004_E', 'TX_RESP_Q004_F', 'TX_RESP_Q004_G', 'TX_RESP_Q004_H', 'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF', 'ID_MUNICIPIO', 
            'DT_ANO_LETIVO_INICIO', 'DT_ANO_LETIVO_TERMINO', 'IN_SITUACAO_CENSO', 'NU_ANO_CENSO', 'CO_UF', 'CO_MUNICIPIO', 'Ideb-2017']'''

'''delete = ['CO_MESORREGIAO', 'CO_MICRORREGIAO', 'CO_DISTRITO', 'TP_LOCALIZACAO',  'CO_ORGAO_REGIONAL', 'TP_DEPENDENCIA',
          'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF', 'ID_MUNICIPIO', 'IN_SITUACAO_CENSO', 'Ideb-2017']'''


#X = dataset.drop(delete, axis=1)

'''X = dataset.drop(['Ideb-2017','IN_SITUACAO_CENSO', 'TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q003_L', 'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D',], axis=1)'''

X = dataset.drop(['Ideb-2017'], axis=1)

In [ ]:
X

ID_PROVA_BRASIL  ID_REGIAO  ...  QtdeTurmasProf  RelAlunProf
28012631             2017          2  ...               8            2
28017846             2017          2  ...              22            5
28017854             2017          2  ...              32            9
28018400             2017          2  ...              34           11
28018435             2017          2  ...              41           11
...                   ...        ...  ...             ...          ...
28009398             2017          2  ...              51           11
28009487             2017          2  ...              13           13
28014839             2017          2  ...              17           11
28010132             2017          2  ...              30           16
28028562             2017          2  ...              22            7

[76 rows x 404 columns]

In [ ]:
filteredColumns = X.dtypes[X.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

#X[listOfColumnNames].fillna(-1, inplace=True)

#X[listOfColumnNames] = X[listOfColumnNames].astype(int)

X[listOfColumnNames].isnull().sum()

Series([], dtype: float64)

In [ ]:
X_transform = minmax(X)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 56.25 %
Naive Bayes: Acurácia do modelo: 75.0 %
Decision Tree: Acurácia do modelo: 56.25 %
SVM: Acurácia do modelo: 50.0 %
Logistic Regression: Acurácia do modelo: 56.25 %
KNN: Acurácia do modelo: 68.75 %


**-Segunda Análise: Aplicando Seleção de Atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_transform, y, 100)

X_embedded

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


ID_MUNICIPIO  ID_AREA  ...  IN_COMUM_EJA_MEDIO  IN_COMUM_EJA_PROF
28012631       2800100        2  ...                   1                  0
28017846       2800308        1  ...                   0                  0
28017854       2800308        1  ...                   0                  0
28018400       2800308        1  ...                   0                  0
28018435       2800308        1  ...                   0                  0
...                ...      ...  ...                 ...                ...
28009398       2807105        2  ...                   0                  0
28009487       2807105        2  ...                   0                  0
28014839       2807204        2  ...                   1                  0
28010132       2807402        2  ...                   0                  0
28028562       2807600        2  ...                   1                  0

[76 rows x 100 columns]

In [ ]:
X_transform_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_transform_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 87.5 %
Naive Bayes: Acurácia do modelo: 81.25 %
Decision Tree: Acurácia do modelo: 62.5 %
SVM: Acurácia do modelo: 81.25 %
Logistic Regression: Acurácia do modelo: 75.0 %
KNN: Acurácia do modelo: 43.75 %


**-Segunda Análise: Aplicando Seleção de Atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_transform, y, 100)

X_wrapper

ID_MUNICIPIO  ID_AREA  ...  IN_COMUM_EJA_FUND  IN_COMUM_EJA_PROF
28012631       2800100        2  ...                  0                  0
28017846       2800308        1  ...                  0                  0
28017854       2800308        1  ...                  0                  0
28018400       2800308        1  ...                  0                  0
28018435       2800308        1  ...                  0                  0
...                ...      ...  ...                ...                ...
28009398       2807105        2  ...                  0                  0
28009487       2807105        2  ...                  0                  0
28014839       2807204        2  ...                  1                  0
28010132       2807402        2  ...                  0                  0
28028562       2807600        2  ...                  0                  0

[76 rows x 100 columns]

In [ ]:
X_transform_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_transform_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 87.5 %
Naive Bayes: Acurácia do modelo: 68.75 %
Decision Tree: Acurácia do modelo: 75.0 %
SVM: Acurácia do modelo: 93.75 %
Logistic Regression: Acurácia do modelo: 75.0 %
KNN: Acurácia do modelo: 62.5 %


**-Segunda Análise: Aplicando Seleção de Atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X, y, 100)

X_selection

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [  0   1   2 153 234 236 239 242 250 251 252 253 254 255 256 261 264 265
 266 267 268 269 275 276 277 278 287 291 292 294 295 306 310 325 341 344
 350 352 353 355 356 357 358 359 360 363 364 367 369 370 371 372 373 374
 375 376 380 381 382 384 386 387 394 398] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


ID_AREA  ID_LOCALIZACAO  ...  QtdeDocentesEfetivos  QtdeTurmasProf
28012631        2               1  ...                     7               8
28017846        1               1  ...                    22              22
28017854        1               1  ...                    32              32
28018400        1               1  ...                    34              34
28018435        1               1  ...                    41              41
...           ...             ...  ...                   ...             ...
28009398        2               1  ...                    46              51
28009487        2               2  ...                    12              13
28014839        2               1  ...                    16              17
28010132        2               1  ...                    27              30
28028562        2               1  ...                    18              22

[76 rows x 100 columns]

In [ ]:
X_transform = minmax(X_selection)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 96.05 %
Naive Bayes: Acurácia do modelo: 78.95 %
Decision Tree: Acurácia do modelo: 92.11 %
SVM: Acurácia do modelo: 93.42 %
Logistic Regression: Acurácia do modelo: 94.74 %
KNN: Acurácia do modelo: 69.74 %


**Segunda Análise: Unindo subconjunto de melhores atributos**

In [ ]:
X_all = pd.concat([X_selection, X_wrapper], axis=1)

X_all = pd.concat([X_all, X_embedded], axis=1)

In [ ]:
#Removendo colunas duplicadas
X_new_all = X_all.T.drop_duplicates().T

In [ ]:
X_transform = minmax(X_new_all)
#X_transform = minmax(X_all)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 94.74 %
Naive Bayes: Acurácia do modelo: 81.58 %
Decision Tree: Acurácia do modelo: 90.79 %
SVM: Acurácia do modelo: 94.74 %
Logistic Regression: Acurácia do modelo: 96.05 %
KNN: Acurácia do modelo: 73.68 %


Resultado com Todos atributos

- Random forest: Acurácia do modelo: 96.05 %
- Naive Bayes: Acurácia do modelo: 82.89 %
- Decision Tree: Acurácia do modelo: 93.42 %
- SVM: Acurácia do modelo: 97.37 %
- Logistic Regression: Acurácia do modelo: 97.37 %
- KNN: Acurácia do modelo: 76.32 %

**Segunda Análise: Aplicando a técnica merge**

Link da Dissertação: https://www.dcc.ufmg.br/pos/cursos/defesas/1930M.PDF


In [ ]:
#Metodo Merge combina os atributos mais frequentes nos melhores conjuntos de seleção 
#Cada atributo irá apresentar um score de acordo com sua presença em um subconjunto de dados
def metodo_merge(selection, embedded, wrapper, ranking):
  for atributo in selection:
    score = ranking[atributo]
    if (atributo in embedded):
      score += 1
      embedded.remove(atributo)
    if (atributo in wrapper):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score

  for atributo in embedded:
    score = ranking[atributo]
    if (atributo in selection):
      score += 1
      embedded.remove(atributo)
    if (atributo in wrapper):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score

  for atributo in wrapper:
    score = ranking[atributo]
    if (atributo in selection):
      score += 1
      embedded.remove(atributo)
    if (atributo in embedded):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score
 
  return ranking      

In [ ]:
#Gerando Listas com os melhores atributos de cada abordagem
selection = list(X_selection.columns)
embedded = list(X_embedded.columns)
wrapper = list(X_wrapper.columns)

#Criando dicionário
values = [1]*300

all = selection + embedded + wrapper

ranking = dict(zip(all, values))

ranking = metodo_merge(selection, embedded, wrapper, ranking)

#Filtrando os atributos que estão em mais de um subconjunto

dic_aux = ranking.items()

filtered = {key:value for key, value in dic_aux if value > 1}

#Gerando X após o merge

X_merge = dataset[list(filtered.keys())]

In [ ]:
X_transform = minmax(X_merge)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 96.05 %
Naive Bayes: Acurácia do modelo: 82.89 %
Decision Tree: Acurácia do modelo: 93.42 %
SVM: Acurácia do modelo: 93.42 %
Logistic Regression: Acurácia do modelo: 93.42 %
KNN: Acurácia do modelo: 76.32 %


In [ ]:
dados_para_exportar = pd.DataFrame(columns=X_merge.columns, data=X_merge.values, index=X_merge.index)
dados_para_exportar.insert(len(dados_para_exportar.columns), "Valor Real Ideb", y.to_numpy())
dados_para_exportar.insert(len(dados_para_exportar.columns), "Previsão Ideb", random_result)


dados_para_exportar

ID_AREA  TX_RESP_Q028_B  ...  Valor Real Ideb    Previsão Ideb
28012631        2               6  ...  abaixo da média  abaixo da média
28017846        1               6  ...   acima da média   acima da média
28017854        1               9  ...  abaixo da média  abaixo da média
28018400        1              14  ...   acima da média   acima da média
28018435        1              20  ...   acima da média   acima da média
...           ...             ...  ...              ...              ...
28009398        2              25  ...   acima da média   acima da média
28009487        2               6  ...   acima da média   acima da média
28014839        2               3  ...   acima da média   acima da média
28010132        2              21  ...   acima da média   acima da média
28028562        2              11  ...   acima da média   acima da média

[76 rows x 93 columns]

In [ ]:
dados_para_exportar = dados_para_exportar.loc[dados_para_exportar['Valor Real Ideb'] ==  dados_para_exportar['Previsão Ideb']]


dados_para_exportar

ID_AREA  TX_RESP_Q028_B  ...  Valor Real Ideb    Previsão Ideb
28012631        2               6  ...  abaixo da média  abaixo da média
28017846        1               6  ...   acima da média   acima da média
28017854        1               9  ...  abaixo da média  abaixo da média
28018400        1              14  ...   acima da média   acima da média
28018435        1              20  ...   acima da média   acima da média
...           ...             ...  ...              ...              ...
28009398        2              25  ...   acima da média   acima da média
28009487        2               6  ...   acima da média   acima da média
28014839        2               3  ...   acima da média   acima da média
28010132        2              21  ...   acima da média   acima da média
28028562        2              11  ...   acima da média   acima da média

[73 rows x 93 columns]

In [ ]:
#dados_para_exportar.to_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/resultado_analise_merge_dataset_completo.csv')
#dados_para_exportar.to_csv('/content/drive/MyDrive/TCC/resultado_analise_merge_dataset_completo.csv')
dados_para_exportar.to_csv('/content/drive/MyDrive/TCC/resultado_analise_merge_dataset_completo_SE.csv')

In [ ]:
dados_para_exportar['TX_RESP_Q005_A'].sum()

210

#### -Terceira análise: variáveis das escolas e professores

In [ ]:
columns_escolas = dataset.columns[237:404]

X_escolas = dataset[columns_escolas]

X_escolas

DT_ANO_LETIVO_INICIO  ...  RelAlunProf
28012631   1486944000000000000  ...            2
28017846   1494201600000000000  ...            5
28017854   1491782400000000000  ...            9
28018400   1492387200000000000  ...           11
28018435   1489363200000000000  ...           11
...                        ...  ...          ...
28009398   1493683200000000000  ...           11
28009487   1490745600000000000  ...           13
28014839   1488412800000000000  ...           11
28010132   1488758400000000000  ...           16
28028562   1487548800000000000  ...            7

[76 rows x 167 columns]

In [ ]:
X_escolas.drop(columns=['DT_ANO_LETIVO_INICIO', 'DT_ANO_LETIVO_TERMINO'], axis=1, inplace=True)

X_transform = minmax(X_escolas)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 43.75 %
Naive Bayes: Acurácia do modelo: 56.25 %
Decision Tree: Acurácia do modelo: 37.5 %
SVM: Acurácia do modelo: 37.5 %
Logistic Regression: Acurácia do modelo: 37.5 %
KNN: Acurácia do modelo: 81.25 %


**Aplicando seleção de atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X_escolas, y, 100)

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [  0   3  11  12  13  14  15  16  17  22  25  26  27  28  29  30  36  37
  38  39  48  52  53  55  56  67  71  86 102 105 111 113 114 116 117 118
 119 120 121 124 125 128 130 131 132 133 134 135 136 137 141 142 143 145
 147 148 155 159] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 50.0 %
Naive Bayes: Acurácia do modelo: 56.25 %
Decision Tree: Acurácia do modelo: 37.5 %
SVM: Acurácia do modelo: 37.5 %
Logistic Regression: Acurácia do modelo: 37.5 %
KNN: Acurácia do modelo: 81.25 %


**Aplicando seleção de atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_transform, y, 100)

X_embedded

ID_REGIAO  ID_AREA  ...  TX_RESP_Q013_D  TX_RESP_Q014_A
28012631          2        2  ...               0               0
28017846          2        1  ...               0               1
28017854          2        1  ...               0               1
28018400          2        1  ...               1               0
28018435          2        1  ...               4               4
...             ...      ...  ...             ...             ...
28009398          2        2  ...               0               0
28009487          2        2  ...               0               2
28014839          2        2  ...               2               1
28010132          2        2  ...               1               1
28028562          2        2  ...               0               0

[76 rows x 70 columns]

In [ ]:
X_transform_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_transform_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 62.5 %
Naive Bayes: Acurácia do modelo: 62.5 %
Decision Tree: Acurácia do modelo: 43.75 %
SVM: Acurácia do modelo: 56.25 %
Logistic Regression: Acurácia do modelo: 62.5 %
KNN: Acurácia do modelo: 56.25 %


**Aplicando seleção de atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_transform, y, 100)

X_wrapper

ID_REGIAO  ID_UF  ...  TX_RESP_Q013_D  TX_RESP_Q014_A
28012631          2     28  ...               0               0
28017846          2     28  ...               0               1
28017854          2     28  ...               0               1
28018400          2     28  ...               1               0
28018435          2     28  ...               4               4
...             ...    ...  ...             ...             ...
28009398          2     28  ...               0               0
28009487          2     28  ...               0               2
28014839          2     28  ...               2               1
28010132          2     28  ...               1               1
28028562          2     28  ...               0               0

[76 rows x 100 columns]

In [ ]:
X_transform_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_transform_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 68.75 %
Naive Bayes: Acurácia do modelo: 56.25 %
Decision Tree: Acurácia do modelo: 68.75 %
SVM: Acurácia do modelo: 68.75 %
Logistic Regression: Acurácia do modelo: 62.5 %
KNN: Acurácia do modelo: 56.25 %
